### PR TITLE
Added Hook method and property -RSTensor

### LINKED ISSUE #177: Replicated Shared Tensor (RST): Hook method and Property
## Feature Description
Write correct functionality of hook method and property behaviour of Replicated Shared Tensor (RST) that has been merged to the codebase. 

## Additional Context
Has a dependency on #164

### PR DESCRIPTION
**Pull request by rasswanth-s**
## Description
Modified the hook method and property of ReplicatedSharedTensor. fixes #177 

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
Added test for hook methods and property.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

**Review by kamathhrishi — CHANGES_REQUESTED** [review]
So the overall structure of ReplicatedSharedTensor is as follows. 

MPCTensor->ReplicatedSharedTensor->List[Torch Tensors]

So that means RSTensor holds shares of torch tensors and not sharetensors. RSTensor and ShareTensors are two seperate types of sharetensors.

File: src/sympc/tensor/replicatedshare_tensor.py [hunk 1]
```diff
@@ -155,13 +157,23 @@ def hook_property(property_name: str) -> Any:
         """
 
         def property_new_share_tensor_getter(_self: "ReplicatedSharedTensor") -> Any:
-            tensor = getattr(_self.tensor, property_name)
-            res = ReplicatedSharedTensor(session=_self.session)
-            res.tensor = tensor
+            tensor1 = getattr(_self.shares[0].tensor, property_name)
+            tensor2 = getattr(_self.shares[1].tensor, property_name)
+            share1 = ShareTensor(session=_self.session)
```
Comment (kamathhrishi) [inline src/sympc/tensor/replicatedshare_tensor.py:162]: RSTensor attribute shares hold a list of torch tensors. The attributes and methods will act upon these torch tensors, not sharetensors. 
Also, parties can hold multiple shares and not necessarily two shares, so it must act on all tensors in the list.

File: src/sympc/tensor/replicatedshare_tensor.py [hunk 2]
```diff
@@ -155,13 +157,23 @@ def hook_property(property_name: str) -> Any:
...
+            share1.tensor = (
+                tensor1  # assign after instance creation to prevent FP encoding.
+            )
+            share2 = ShareTensor(session=_self.session)
```
Comment (kamathhrishi) [inline src/sympc/tensor/replicatedshare_tensor.py:166]: Same here.

File: src/sympc/tensor/replicatedshare_tensor.py [hunk 3]
```diff
@@ -192,16 +204,25 @@ def hook_method(method_name: str) -> Callable[..., Any]:
         def method_new_rs_tensor(
             _self: "ReplicatedSharedTensor", *args: List[Any], **kwargs: Dict[Any, Any]
         ) -> Any:
-            method = getattr(_self.tensor, method_name)
-            tensor = method(*args, **kwargs)
-            res = ReplicatedSharedTensor(session=_self.session, shares=_self.shares)
-            res.tensor = tensor
+            tensor1 = getattr(_self.shares[0].tensor, method_name)(*args, **kwargs)
+            tensor2 = getattr(_self.shares[1].tensor, method_name)(*args, **kwargs)
+            share1 = ShareTensor(session=_self.session)
```
Comment (kamathhrishi) [inline src/sympc/tensor/replicatedshare_tensor.py:209]: Same as attributes

File: tests/sympc/tensor/replicatedshare_tensor_test.py [hunk 4]
```diff
@@ -1,6 +1,48 @@
...
+
+
+def test_hook_property(get_clients) -> None:
+    alice, bob = get_clients(2)
+    session = Session(parties=[alice, bob])
+    SessionManager.setup_mpc(session)
+
+    x = torch.randn(1, 3)
+    y = torch.randn(1, 3)  # noqa: F841
+    stx = ShareTensor(data=x, session=session)
```
Comment (kamathhrishi) [inline tests/sympc/tensor/replicatedshare_tensor_test.py:42]: The tests will have to modify according to the above behavior.

**Review by kamathhrishi — CHANGES_REQUESTED** [review]

**Review by gmuraru — COMMENTED** [review]
We have at the moment:
MPCTensor holds ShareTensor holds Torch Tensors

For Falcon it might be
MPCTensor holds RSS Tensor hold Torch Tensor

File: src/sympc/tensor/replicatedshare_tensor.py (hunk 1 above)
Comment (rasswanth-s) [reply src/sympc/tensor/replicatedshare_tensor.py:162]: It actually applies to all the tensors.Given a secret x  ,each party holds two shares. Applying the hook method or property to an MPCTensor iterates through all the RSTensor and applied the function. Correct me if I am wrong.

**Comment by rasswanth-s** [comment]
> We have at the moment:
> MPCTensor holds ShareTensor holds Torch Tensors
> 
> For Falcon it might be
> MPCTensor holds RSS Tensor hold Torch Tensor

The init method had List[ShareTensor], I got confused ,I would refactor it

File: src/sympc/tensor/replicatedshare_tensor.py [hunk 5]
```diff
@@ -192,20 +209,24 @@ def hook_method(method_name: str) -> Callable[..., Any]:
         def method_new_rs_tensor(
             _self: "ReplicatedSharedTensor", *args: List[Any], **kwargs: Dict[Any, Any]
         ) -> Any:
-            method = getattr(_self.tensor, method_name)
-            tensor = method(*args, **kwargs)
-            res = ReplicatedSharedTensor(session=_self.session, shares=_self.shares)
-            res.tensor = tensor
+            shares = []
+            nr_parties = _self.session.nr_parties
+            for i in range(nr_parties - 1):  # each party has n-1 shares
```
Comment (kamathhrishi) [inline src/sympc/tensor/replicatedshare_tensor.py:214]: Same here.

File: src/sympc/tensor/replicatedshare_tensor.py [hunk 6]
```diff
@@ -154,18 +165,24 @@ def hook_property(property_name: str) -> Any:
...
 
-        def property_new_share_tensor_getter(_self: "ReplicatedSharedTensor") -> Any:
-            tensor = getattr(_self.tensor, property_name)
-            res = ReplicatedSharedTensor(session=_self.session)
-            res.tensor = tensor
+        def property_new_rs_tensor_getter(_self: "ReplicatedSharedTensor") -> Any:
+            shares = []
+            nr_parties = _self.session.nr_parties
+
+            for i in range(nr_parties - 1):  # each party has n-1 shares.
```
[the remark this answers is not in the document]
Comment (rasswanth-s) [reply src/sympc/tensor/replicatedshare_tensor.py:172]: Sure @kamathhrishi

File: src/sympc/tensor/replicatedshare_tensor.py [hunk 7]
```diff
@@ -30,11 +37,15 @@ class ReplicatedSharedTensor(metaclass=SyMPCTensor):
...
+    def __init__(
+        self,
+        shares: Optional[Union[float, int, torch.Tensor]] = None,
+        session: Optional[Session] = None,
+    ):
         """Initialize ShareTensor.
 
         Args:
-            shares (Optional[List[ShareTensor]]): Shares from which RSTensor is created.
+            shares (Optional[Union[float,int,torch.Tensor]]): Shares from which RSTensor is created.
```
[the remark this answers is not in the document]
Comment (rasswanth-s) [reply src/sympc/tensor/replicatedshare_tensor.py:48]: @kamathhrishi , I think it should be List[Union[float,int,torch.Tensor]].

[the remark this answers is not in the document]
Comment (kamathhrishi) [reply src/sympc/tensor/replicatedshare_tensor.py:48]: Yeah that is it.

File: tests/sympc/tensor/replicatedshare_tensor_test.py [hunk 8]
```diff
@@ -1,6 +1,43 @@
...
 
 
-def test_import_RSTensor():
+def test_import_RSTensor() -> None:
 
     ReplicatedSharedTensor()
+
+
+def test_hook_method(get_clients) -> None:
+    alice, bob = get_clients(2)
```
Comment (aanurraj) [inline tests/sympc/tensor/replicatedshare_tensor_test.py:15]: Q: We are implementing RST mainly for FALCON don't you think we should check it for 3 parties?
After all, Falcon is A 3-Party Machine Learning Service. :D

**Review by aanurraj — CHANGES_REQUESTED** [review]

File: tests/sympc/tensor/replicatedshare_tensor_test.py (hunk 8 above)
Comment (rasswanth-s) [reply tests/sympc/tensor/replicatedshare_tensor_test.py:15]: Yeah you are right :slightly_smiling_face:,I thought to include the sanity check in protocol in later commits.

File: src/sympc/tensor/replicatedshare_tensor.py [hunk 9]
```diff
@@ -5,12 +5,19 @@
...
+import torch
+
+from sympc.session import Session
 
 from .tensor import SyMPCTensor
 
-PROPERTIES_NEW_SHARE_TENSOR: Set[str] = {"T"}
-METHODS_NEW_SHARE_TENSOR: Set[str] = {"unsqueeze", "view", "t", "sum", "clone"}
+PROPERTIES_NEW_RS_TENSOR: Set[str] = {"T"}
+METHODS_NEW_RS_TENSOR: Set[str] = {"unsqueeze", "view", "t", "sum", "clone"}
```
[the remark this answers is not in the document]
Comment (rasswanth-s) [reply src/sympc/tensor/replicatedshare_tensor.py:20]: Then I think we can defer the PR,merge it when RSTensor functionality is completely implemented.

File: tests/sympc/tensor/replicatedshare_tensor_test.py [hunk 10]
```diff
@@ -1,6 +1,43 @@
...
+
+    assert rst.numel() == x.numel()
+    assert (rst.t().shares[0] == x.t()).all()
+    assert (rst.unsqueeze(dim=0).shares[0] == x.unsqueeze(dim=0)).all()
+    assert (rst.view(3, 1).shares[0] == x.view(3, 1)).all()
+    assert (rst.sum().shares[0] == x.sum()).all()
+
+
+def test_hook_property(get_clients) -> None:
+    alice, bob, charlie = get_clients(3)
```
Comment (gmuraru) [inline tests/sympc/tensor/replicatedshare_tensor_test.py:33]: NITPICK: Use
```clients = get_clients(3)```
and directly pass them to session as ```Session(parties=clients)```

File: tests/sympc/tensor/replicatedshare_tensor_test.py [hunk 11]
```diff
@@ -1,6 +1,43 @@
...
+
+    x = torch.randn(1, 3)
+    y = torch.randn(1, 3)  # noqa: F841
+    shares = [x, y]
+
+    rst = ReplicatedSharedTensor(shares=shares, session=session)
+
+    assert rst.numel() == x.numel()
+    assert (rst.t().shares[0] == x.t()).all()
+    assert (rst.unsqueeze(dim=0).shares[0] == x.unsqueeze(dim=0)).all()
```
Comment (gmuraru) [inline tests/sympc/tensor/replicatedshare_tensor_test.py:27]: Check also for y.

It might be a good idea to have a testing function (even if there will be a lot of duplicate code - for each forward method/property).

Comment (rasswanth-s) [reply tests/sympc/tensor/replicatedshare_tensor_test.py:27]: Sure @gmuraru

**Review by kamathhrishi — APPROVED** [review]

**Comment by kamathhrishi** [comment]
Looks good to me.  @gmuraru and @aanurraj Have a look at it.

**Review by gmuraru — APPROVED** [review]
LGTM! @aanurraj could you have a look?

**Review by aanurraj — APPROVED** [review]
